pkg/volume/gce_pd/attacher.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce_pd
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/keymutex"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+type gcePersistentDiskAttacher struct {
+	host volume.VolumeHost
+}
+
+var _ volume.Attacher = &gcePersistentDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &gcePersistentDiskPlugin{}
+
+// Singleton key mutex for keeping attach/detach operations for the
+// same PD atomic
+// TODO(swagiaal): Once the Mount/Unmount manager is implemented this
+// will no longer be needed and should be removed.
+var attachDetachMutex = keymutex.NewKeyMutex()
+
+func (plugin *gcePersistentDiskPlugin) NewAttacher() (volume.Attacher, error) {
+	return &gcePersistentDiskAttacher{host: plugin.host}, nil
+}
+
+func (attacher *gcePersistentDiskAttacher) Attach(spec *volume.Spec, hostName string) error {
+	volumeSource, readOnly := getVolumeSource(spec)
+	pdName := volumeSource.PDName
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(pdName)
+	defer attachDetachMutex.UnlockKey(pdName)
+
+	gceCloud, err := getCloudProvider(attacher.host.GetCloudProvider())
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if numRetries > 0 {
+			glog.Warningf("Retrying attach for GCE PD %q (retry count=%v).", pdName, numRetries)
+		}
+
+		if err = gceCloud.AttachDisk(pdName, hostName, readOnly); err != nil {
+			glog.Errorf("Error attaching PD %q: %+v", pdName, err)
+			time.Sleep(errorSleepDuration)
+			continue
+		}
+
+		return nil
+	}
+
+	return err
+}
+
+func (attacher *gcePersistentDiskAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
```

Comment:
I agree. But for slightly different reasons. Both could require cloud provider--attach for attaching, wait for attach because some plugins like AWS EBS, waitForAttach would need to first call the cloud provider to figure out where on the node to expect the volume (it is not deterministic). But attach would never require the kubelet specific volume host stuff where as waitForAttach would.